# DS-SF-34 | 01 | What is Data Science | Assignment | Starter Code

## Python Review

Programming is a must-have skill for data scientists.  Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [29]:
# TODO
def sum_multiple_3_or_5(n,m):
    sum = 0
    for i in range(n,m):
        if (i % 3 == 0) or (i % 5 == 0):
            sum = sum + 1
            print (i)
            
    return i

sum_multiple_3_or_5(1,1000)

        


        

3
5
6
9
10
12
15
18
20
21
24
25
27
30
33
35
36
39
40
42
45
48
50
51
54
55
57
60
63
65
66
69
70
72
75
78
80
81
84
85
87
90
93
95
96
99
100
102
105
108
110
111
114
115
117
120
123
125
126
129
130
132
135
138
140
141
144
145
147
150
153
155
156
159
160
162
165
168
170
171
174
175
177
180
183
185
186
189
190
192
195
198
200
201
204
205
207
210
213
215
216
219
220
222
225
228
230
231
234
235
237
240
243
245
246
249
250
252
255
258
260
261
264
265
267
270
273
275
276
279
280
282
285
288
290
291
294
295
297
300
303
305
306
309
310
312
315
318
320
321
324
325
327
330
333
335
336
339
340
342
345
348
350
351
354
355
357
360
363
365
366
369
370
372
375
378
380
381
384
385
387
390
393
395
396
399
400
402
405
408
410
411
414
415
417
420
423
425
426
429
430
432
435
438
440
441
444
445
447
450
453
455
456
459
460
462
465
468
470
471
474
475
477
480
483
485
486
489
490
492
495
498
500
501
504
505
507
510
513
515
516
519
520
522
525
528
530
531
534
535
537
540
543
545
546
549
550
552
555
558
560
561
56

999

Answer: TODO

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + n}{2u_m}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [38]:

def sqrt(m):
    u = 1. 
    while True:
        v = (u ** 2 + m) / (2 * u) 
        if abs(u / v - 1) < 10 ** -12:
            break
        u = v 
    return v



Answer: TODO

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [43]:
def is_prime(n):
    for k in range (2, n):
        if n % k == 0:
            return false
    return True    
    
    
sum = 0
for i in range(2, 1000):
    if is_prime(i):
        sum += i
sum        
    

NameError: name 'false' is not defined

Answer: TODO

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [ ]:
# TODO

Answer: TODO

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [ ]:
# TODO

Answer: TODO

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [ ]:
# TODO

Answer: TODO

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [ ]:
# TODO

Answer: TODO

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [ ]:
# TODO

Answer: TODO